### ISSUE TITLE
Regression: insidePolygon returns wrong result 

### ISSUE THREAD

**Issue by kfigiela** [noinia/hgeometry#14]
Update of hgeometry causes one of my internal unit tests to fail. It looks like there was regression introduced between 48bf3a3fcd1519c2a46ed8b9ef7ea70f7197bb4d (last correct result) and 67652270a96eccacadec315cb5526b85c4f305ab (this and later break the test). 5dde8bdaede07d8d5204ce7f19917fc4d81d7424 in between does not build. 

I have isolated data from internal unit test as a minimal example: https://github.com/restaumatic/hgeometry/blob/2c25f44be34cd34a2e59c88f0406e37926fb77e5/test/Data/Geometry/Polygon/RegressionSpec.hs

**Comment by noinia** [comment]
Thanks for reporting! Seems to be some floating point issue ([see](https://github.com/noinia/hgeometry/blob/pointInPolyRegression/test/Data/Geometry/Polygon/RegressionSpec.hs) ; using Rational or Pico as numeric type (instead of Double) results in correct output). I'll see if I can update my code s.t. the testcases also pass with Double, but IMO the real fix is to use a better numeric type; Double just sucks for geometric computations.

**Comment by noinia** [comment]
(note to self: after some experimenting I think [this definition of onSubLine2UB]((https://github.com/noinia/hgeometry/commit/5dde8bdaede07d8d5204ce7f19917fc4d81d7424#diff-309f8b01ed16aa1ad41387191b153b77L98) is the culprit; toOffset does not work very well for doubles. I'll try to whip up a better definition (similar to onSubLine2), but it is getting somewhat ugly when 'r' may be UnBounded :(

**Comment by kfigiela** [comment]
Thanks for prompt response! Will investigate use of other data type for numbers tomorrow. Luckily the place we use this is pretty isolated so it should not be a big change.

**Comment by kfigiela** [comment]
I switched to `Rational` and it works like a charm. Thank you @noinia again for fast response. I have not anticipated to stumble upon floating point issue here.

**Comment by Lemmih** [comment]
Improving `onSubLine2` is not enough. We need to change `inPolygon` such that it merely counts intersections instead of finding the intersection points. We can tell if a line (or half-line) intersects a line segment even if we don't have enough precision to represent an intersection point that satisfies `onSegment2`.

Once we separate finding intersection points from testing if intersections exist, then this issue can be solved.

**Comment by Lemmih** [comment]
I was wrong! SubLine intersection is now much more robust and the test case no longer fails. The intersection code was written with the assumption that all segment intersections would lie on the line between the two points that define the segment. This is true when you have arbitrary precision but unfortunately not true for Doubles. Fortunately, we don't need it to be true and a slight refactoring of the code has relaxed this restriction.

TL;DR: Fixed in #106. Will close this issue when the PR is merged.

**Comment by Lemmih** [comment]
#106 contains both a test-case and a fix. Closing.
